File: modules/native-audio/ios/AccentClassification.swift
import Foundation

/// Controls accent behavior for subdivision pulses only.
/// Independent from the beat-level accent pattern.
enum SubdivisionAccentMode {
  case off
  case groupStart
  case everyNth
  case custom
}

/// Live audible roles. `.subdivision` is retained for compatibility only (unused for ticks).
enum ClickSoundKind {
  case bar
  case accent
  case click
  case subdivision
}

enum AccentClassification {
  private static let defaultSubdivisionAccentMode: SubdivisionAccentMode = .off
  private static let defaultSubdivisionAccentEveryNth = 4
  private static let defaultSubdivisionAccentPattern: [Bool] = []
  private static let defaultBarStartEnabled = true

  static func resolveTickAccent(
    beatIndexInBar: Int,
    subdivisionIndex: Int,
    accentPattern: [Bool],
    ticksPerBeat: Int,
    subdivisionAccentMode: SubdivisionAccentMode = defaultSubdivisionAccentMode,
    subdivisionAccentEveryNth: Int = defaultSubdivisionAccentEveryNth,
    subdivisionAccentPattern: [Bool] = defaultSubdivisionAccentPattern
  ) -> Bool {
    let beatIsAccented = resolveBeatAccent(beatIndexInBar: beatIndexInBar, accentPattern: accentPattern)
    let subdivisionIsAccented = resolveSubdivisionAccent(
      beatIndexInBar: beatIndexInBar,
      subdivisionIndex: subdivisionIndex,
      ticksPerBeat: ticksPerBeat,
      subdivisionAccentMode: subdivisionAccentMode,
      subdivisionAccentEveryNth: subdivisionAccentEveryNth,
      subdivisionAccentPattern: subdivisionAccentPattern,
      beatIsAccented: beatIsAccented
    )

    return subdivisionIsAccented || (beatIsAccented && subdivisionIndex == 0)
  }

  /// Priority: Bar Start (downbeat only, when enabled) → accent logic → Click.
  /// Quick Metronome: beat-accent pattern does not apply to beat 1 (Bar Start owns that slot).
  /// Subdivision accents still apply on every beat, including beat 1.
  static func resolveClickSoundKind(
    beatIndexInBar: Int,
    subdivisionIndex: Int,
    accentPattern: [Bool],
    ticksPerBeat: Int,
    subdivisionAccentMode: SubdivisionAccentMode = defaultSubdivisionAccentMode,
    subdivisionAccentEveryNth: Int = defaultSubdivisionAccentEveryNth,
    subdivisionAccentPattern: [Bool] = defaultSubdivisionAccentPattern,
    barStartEnabled: Bool = defaultBarStartEnabled
  ) -> ClickSoundKind {
    if barStartEnabled && isBarStartHit(
      beatIndexInBar: beatIndexInBar,
      subdivisionIndex: subdivisionIndex
    ) {
      return .bar
    }

    let beatIsAccented = resolveBeatAccent(beatIndexInBar: beatIndexInBar, accentPattern: accentPattern)
    // Beat 1 main-beat accent is owned by Bar Start; do not fall through to accentPattern[0].
    let beatAccentForHit = beatIndexInBar != 0 && beatIsAccented

    if isBeatAccentHit(
      beatIsAccented: beatAccentForHit,
      subdivisionIndex: subdivisionIndex,
      ticksPerBeat: ticksPerBeat
    ) {
      return .accent
    }

    if resolveSubdivisionAccent(
      beatIndexInBar: beatIndexInBar,
      subdivisionIndex: subdivisionIndex,
      ticksPerBeat: ticksPerBeat,
      subdivisionAccentMode: subdivisionAccentMode,
      subdivisionAccentEveryNth: subdivisionAccentEveryNth,
      subdivisionAccentPattern: subdivisionAccentPattern,
      beatIsAccented: beatIsAccented
    ) {
      return .accent
    }

    return .click
  }

  /// Song timeline: accent from compiled event; Bar Start only overrides the downbeat when enabled.
  /// Disabling Bar Start removes only BAR — compiled accent on beat 1 still applies.
  static func resolveClickSoundKindFromTickAccent(
    isAccent: Bool,
    beatIndexInBar: Int,
    subdivisionIndex: Int,
    barStartEnabled: Bool = defaultBarStartEnabled
  ) -> ClickSoundKind {
    if barStartEnabled && isBarStartHit(
      beatIndexInBar: beatIndexInBar,
      subdivisionIndex: subdivisionIndex
    ) {
      return .bar
    }

    if isAccent {
      return .accent
    }

    return .click
  }

  static func resolveBeatAccent(beatIndexInBar: Int, accentPattern: [Bool]) -> Bool {
    if accentPattern.isEmpty {
      return false
    }

    return accentPattern[beatIndexInBar % accentPattern.count]
  }

  static func resolveSubdivisionAccent(
    beatIndexInBar: Int,
    subdivisionIndex: Int,
    ticksPerBeat: Int,
    subdivisionAccentMode: SubdivisionAccentMode,
    subdivisionAccentEveryNth: Int,
    subdivisionAccentPattern: [Bool],
    beatIsAccented: Bool
  ) -> Bool {
    if ticksPerBeat <= 1 || subdivisionIndex < 0 || subdivisionIndex >= ticksPerBeat {
      return false
    }

    switch subdivisionAccentMode {
    case .off:
      return false
    case .groupStart:
      return subdivisionIndex == 0
    case .everyNth:
      if subdivisionAccentEveryNth <= 0 {
        return false
      }

      let globalIndex = beatIndexInBar * ticksPerBeat + subdivisionIndex
      return globalIndex % subdivisionAccentEveryNth == 0
    case .custom:
      return resolveCustomSubdivisionAccent(
        subdivisionIndex: subdivisionIndex,
        pattern: subdivisionAccentPattern
      )
    }
  }

  static func resolveCustomSubdivisionAccent(
    subdivisionIndex: Int,
    pattern: [Bool]
  ) -> Bool {
    if pattern.isEmpty {
      return false
    }

    return pattern[subdivisionIndex % pattern.count]
  }

  private static func isBarStartHit(
    beatIndexInBar: Int,
    subdivisionIndex: Int
  ) -> Bool {
    beatIndexInBar == 0 && subdivisionIndex == 0
  }

  private static func isBeatAccentHit(
    beatIsAccented: Bool,
    subdivisionIndex: Int,
    ticksPerBeat: Int
  ) -> Bool {
    if !beatIsAccented {
      return false
    }

    if ticksPerBeat <= 1 {
      return true
    }

    return subdivisionIndex == 0
  }
}

File: modules/native-audio/android/src/main/java/expo/modules/nativeaudio/AccentClassification.kt
package expo.modules.nativeaudio

/**
 * Controls accent behavior for subdivision pulses only.
 * Independent from the beat-level accent pattern.
 */
internal enum class SubdivisionAccentMode {
  OFF,
  GROUP_START,
  EVERY_NTH,
  CUSTOM,
}

/**
 * Live audible roles. [SUBDIVISION] is retained for compatibility only (unused for ticks).
 */
internal enum class ClickSoundKind {
  BAR,
  ACCENT,
  CLICK,
  SUBDIVISION,
}

internal object AccentClassification {
  private val defaultSubdivisionAccentMode = SubdivisionAccentMode.OFF
  private const val defaultSubdivisionAccentEveryNth = 4
  private val defaultSubdivisionAccentPattern = booleanArrayOf()
  private const val defaultBarStartEnabled = true

  fun resolveTickAccent(
    beatIndexInBar: Int,
    subdivisionIndex: Int,
    accentPattern: BooleanArray,
    ticksPerBeat: Int,
    subdivisionAccentMode: SubdivisionAccentMode = defaultSubdivisionAccentMode,
    subdivisionAccentEveryNth: Int = defaultSubdivisionAccentEveryNth,
    subdivisionAccentPattern: BooleanArray = defaultSubdivisionAccentPattern,
  ): Boolean {
    val beatIsAccented = resolveBeatAccent(beatIndexInBar, accentPattern)
    val subdivisionIsAccented = resolveSubdivisionAccent(
      beatIndexInBar,
      subdivisionIndex,
      ticksPerBeat,
      subdivisionAccentMode,
      subdivisionAccentEveryNth,
      subdivisionAccentPattern,
      beatIsAccented,
    )

    return subdivisionIsAccented || (beatIsAccented && subdivisionIndex == 0)
  }

  /**
   * Priority: Bar Start (downbeat only, when enabled) → accent logic → Click.
   * Quick Metronome: beat-accent pattern does not apply to beat 1 (Bar Start owns that slot).
   * Subdivision accents still apply on every beat, including beat 1.
   */
  fun resolveClickSoundKind(
    beatIndexInBar: Int,
    subdivisionIndex: Int,
    accentPattern: BooleanArray,
    ticksPerBeat: Int,
    subdivisionAccentMode: SubdivisionAccentMode = defaultSubdivisionAccentMode,
    subdivisionAccentEveryNth: Int = defaultSubdivisionAccentEveryNth,
    subdivisionAccentPattern: BooleanArray = defaultSubdivisionAccentPattern,
    barStartEnabled: Boolean = defaultBarStartEnabled,
  ): ClickSoundKind {
    if (barStartEnabled && isBarStartHit(beatIndexInBar, subdivisionIndex)) {
      return ClickSoundKind.BAR
    }

    val beatIsAccented = resolveBeatAccent(beatIndexInBar, accentPattern)
    // Beat 1 main-beat accent is owned by Bar Start; do not fall through to accentPattern[0].
    val beatAccentForHit = beatIndexInBar != 0 && beatIsAccented

    if (isBeatAccentHit(beatAccentForHit, subdivisionIndex, ticksPerBeat)) {
      return ClickSoundKind.ACCENT
    }

    if (
      resolveSubdivisionAccent(
        beatIndexInBar,
        subdivisionIndex,
        ticksPerBeat,
        subdivisionAccentMode,
        subdivisionAccentEveryNth,
        subdivisionAccentPattern,
        beatIsAccented,
      )
    ) {
      return ClickSoundKind.ACCENT
    }

    return ClickSoundKind.CLICK
  }

  /**
   * Song timeline: accent from compiled event; Bar Start only overrides the downbeat when enabled.
   * Disabling Bar Start removes only BAR — compiled accent on beat 1 still applies.
   */
  fun resolveClickSoundKindFromTickAccent(
    isAccent: Boolean,
    beatIndexInBar: Int,
    subdivisionIndex: Int,
    barStartEnabled: Boolean = defaultBarStartEnabled,
  ): ClickSoundKind {
    if (barStartEnabled && isBarStartHit(beatIndexInBar, subdivisionIndex)) {
      return ClickSoundKind.BAR
    }

    if (isAccent) {
      return ClickSoundKind.ACCENT
    }

    return ClickSoundKind.CLICK
  }

  fun resolveBeatAccent(beatIndexInBar: Int, accentPattern: BooleanArray): Boolean {
    if (accentPattern.isEmpty()) {
      return false
    }

    return accentPattern[beatIndexInBar % accentPattern.size]
  }

  fun resolveSubdivisionAccent(
    beatIndexInBar: Int,
    subdivisionIndex: Int,
    ticksPerBeat: Int,
    subdivisionAccentMode: SubdivisionAccentMode,
    subdivisionAccentEveryNth: Int,
    subdivisionAccentPattern: BooleanArray,
    beatIsAccented: Boolean,
  ): Boolean {
    if (ticksPerBeat <= 1 || subdivisionIndex < 0 || subdivisionIndex >= ticksPerBeat) {
      return false
    }

    return when (subdivisionAccentMode) {
      SubdivisionAccentMode.OFF -> false
      SubdivisionAccentMode.GROUP_START -> subdivisionIndex == 0
      SubdivisionAccentMode.EVERY_NTH -> {
        if (subdivisionAccentEveryNth <= 0) {
          false
        } else {
          val globalIndex = beatIndexInBar * ticksPerBeat + subdivisionIndex
          globalIndex % subdivisionAccentEveryNth == 0
        }
      }
      SubdivisionAccentMode.CUSTOM -> resolveCustomSubdivisionAccent(
        subdivisionIndex,
        subdivisionAccentPattern,
      )
    }
  }

  fun resolveCustomSubdivisionAccent(
    subdivisionIndex: Int,
    pattern: BooleanArray,
  ): Boolean {
    if (pattern.isEmpty()) {
      return false
    }

    return pattern[subdivisionIndex % pattern.size]
  }

  private fun isBarStartHit(
    beatIndexInBar: Int,
    subdivisionIndex: Int,
  ): Boolean {
    return beatIndexInBar == 0 && subdivisionIndex == 0
  }

  private fun isBeatAccentHit(
    beatIsAccented: Boolean,
    subdivisionIndex: Int,
    ticksPerBeat: Int,
  ): Boolean {
    if (!beatIsAccented) {
      return false
    }

    if (ticksPerBeat <= 1) {
      return true
    }

    return subdivisionIndex == 0
  }
}
